src/main/java/com/apicatalog/jsonld/flattening/NodeMapBuilder.java

```diff
@@ -123,296 +167,347 @@ public NodeMap build() throws JsonLdError {
+        elementObject.remove(Keywords.INDEX);
+    }
 
-                id = ((JsonString)idValue).getString();
+    private void handle6_5(String id) {
+        // 6.5.1.
+        if (nodeMap.contains(activeGraph, id, activeProperty)) {
 
-                if (BlankNode.hasPrefix(id)) {
-                    id = nodeMap.createIdentifier(id);
-                }
-                elementObject.remove(Keywords.ID);
+            final JsonArray activePropertyValue = nodeMap.get(activeGraph, id, activeProperty).asJsonArray();
 
-            // 6.2.
-            } else {
-                id = nodeMap.createIdentifier();
+            if (activePropertyValue.stream().filter(JsonUtils::isObject).noneMatch(e -> Objects.equals(e.asJsonObject(), JsonUtils.toJsonObject(referencedNode)))) {
+                nodeMap.set(activeGraph, id, activeProperty, JsonProvider.instance().createArrayBuilder(activePropertyValue).add(JsonUtils.toJsonObject(referencedNode)).build());
             }
 
-            // 6.3.
-            if (id != null && !nodeMap.contains(activeGraph, id)) {
-                nodeMap.set(activeGraph, id, Keywords.ID, JsonProvider.instance().createValue(id));
-            }
+            // 6.5.2.
+        } else {
+            nodeMap.set(activeGraph, id, activeProperty, JsonProvider.instance().createArrayBuilder().add(JsonUtils.toJsonObject(referencedNode)).build());
+        }
+    }
 
-            // 6.4.
+    private void handle6_6(String id) {
+        // 6.6.1.
+        final JsonObject reference = JsonProvider.instance().createObjectBuilder().add(Keywords.ID, id).build();
 
-            // 6.5.
-            if (referencedNode != null) {
+        // 6.6.2.
+        if (list == null) {
 
-                // 6.5.1.
-                if (nodeMap.contains(activeGraph, id, activeProperty)) {
+            // 6.6.2.2.
+            if (nodeMap.contains(activeGraph, activeSubject, activeProperty)) {
 
-                    final JsonArray activePropertyValue = nodeMap.get(activeGraph, id, activeProperty).asJsonArray();
+                final JsonArray activePropertyValue = nodeMap.get(activeGraph, activeSubject, activeProperty).asJsonArray();
 
-                    if (activePropertyValue.stream().filter(JsonUtils::isObject).noneMatch(e -> Objects.equals(e.asJsonObject(), JsonUtils.toJsonObject(referencedNode)))) {
-                        nodeMap.set(activeGraph, id, activeProperty, JsonProvider.instance().createArrayBuilder(activePropertyValue).add(JsonUtils.toJsonObject(referencedNode)).build());
+                if (noneMatch(activePropertyValue, reference)) {
+                    JsonArray build;
+                    if (activePropertyValue.isEmpty()) {
+                        build = JsonProvider.instance().createArrayBuilder(List.of(reference)).build();
+                    } else {
+                        build = JsonProvider.instance().createArrayBuilder(activePropertyValue).add(reference).build();
```

Comment:
This is currently the most resource intensive part of the code. I'm not sure what to do about it since it's more of an architectural issue.

Essentially the problem is that every time we add the `reference` object we need to create a copy of the JsonArray object. 

This code allocates 500 MB worth of JsonArray objects when converting the benchmark file containing 600 000 triples. This is 1/3 of all the allocations.